File: scripts_n_helpers/printEndOfMakeInfo.mjs
#!/usr/bin/env node

import { 
  STSClient, 
  GetCallerIdentityCommand 
} from "@aws-sdk/client-sts";
import { 
  CognitoIdentityProviderClient, 
  AdminCreateUserCommand,
  AdminAddUserToGroupCommand,
  AdminSetUserPasswordCommand,
  UsernameExistsException
} from "@aws-sdk/client-cognito-identity-provider";
import { promises as fs } from 'fs';
import chalk from 'chalk';
import readline from 'readline';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

// Parse command line arguments
const argv = yargs(hideBin(process.argv))
  .option('email', {
      alias: 'e',
      description: 'Email address for the user',
      type: 'string'
  })
  .parse();

function generateRandomPassword() {
  const lowercase = 'abcdefghijklmnopqrstuvwxyz';
  const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  const numbers = '0123456789';
  const specialChars = '#$%&()*+,-./';

  const generatePart = (chars, length) => {
      let result = '';
      for (let i = 0; i < length; i++) {
          result += chars.charAt(Math.floor(Math.random() * chars.length));
      }
      return result;
  };

  const part1 = generatePart(lowercase + numbers, 8);
  const part2 = generatePart(uppercase + numbers, 8);
  const specialChar = specialChars.charAt(Math.floor(Math.random() * specialChars.length));

  return part1 + specialChar + part2;
}

async function getUserInput(prompt) {
  const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
  });

  return new Promise(resolve => {
      rl.question(prompt, (answer) => {
          rl.close();
          resolve(answer.trim());
      });
  });
}

async function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
  // Get email address
  const emailAddress = argv.email || await getUserInput("Please enter your email address: ");

  // Generate password
  const password = generateRandomPassword();
  const passwordOutput = `Your randomly generated password is: ${chalk.yellow(password)}`;

  // Initialize AWS clients
  const region = 'us-west-2';
  const stsClient = new STSClient({ region });
  const cognitoClient = new CognitoIdentityProviderClient({ region });

  try {
      // Get AWS account ID
      const stsResponse = await stsClient.send(new GetCallerIdentityCommand({}));
      const accountId = stsResponse.Account;

      // Read outputs file
      const outputs = JSON.parse(await fs.readFile('./outputs.json', 'utf8'));

      for (const [key, value] of Object.entries(outputs)) {
          if (key.includes("-backend-")) continue;
          if (key.includes("-")) {
              for (const [innerKey, innerValue] of Object.entries(value)) {
                  if (innerKey.startsWith("WebsiteUserInterfaceDomainName")) {
                      console.log(`DeepRacer Chatbot URL: ${chalk.blue(innerValue)}`);
                  } else if (innerKey.startsWith("ExportsOutputRefAuthenticationUserPool")) {
                      const userPoolId = innerValue;

                      console.log("\nStarting user creation process...");

                      // Step 1: Create user
                      console.log("\nExecuting: Creating user");
                      try {
                          await cognitoClient.send(new AdminCreateUserCommand({
                              UserPoolId: userPoolId,
                              Username: emailAddress,
                              MessageAction: 'SUPPRESS'
                          }));
                          console.log(chalk.green("Success!"));
                      } catch (error) {
                          if (error instanceof UsernameExistsException) {
                              console.log(chalk.yellow("Note: User already exists"));
                          } else {
                              console.log(chalk.red(`✗ Error creating user: ${error.message}`));
                              process.exit(1);
                          }
                      }

                      await sleep(2000);

                      // Step 2: Add user to group
                      console.log("\nExecuting: Adding user to group");
                      try {
                          await cognitoClient.send(new AdminAddUserToGroupCommand({
                              UserPoolId: userPoolId,
                              Username: emailAddress,
                              GroupName: 'Users'
                          }));
                          console.log(chalk.green("Success!"));
                      } catch (error) {
                          console.log(chalk.red(`✗ Error adding user to group: ${error.message}`));
                          process.exit(1);
                      }

                      await sleep(1000);

                      // Step 3: Set permanent password
                      console.log("\nExecuting: Setting user password");
                      try {
                          await cognitoClient.send(new AdminSetUserPasswordCommand({
                              UserPoolId: userPoolId,
                              Username: emailAddress,
                              Password: password,
                              Permanent: true
                          }));
                          console.log(chalk.green("Success!"));
                          console.log('');
                          console.log(passwordOutput);
                      } catch (error) {
                          console.log(chalk.red(`✗ Error setting password: ${error.message}`));
                          process.exit(1);
                      }
                  }
              }
          }
      }
  } catch (error) {
      console.error(chalk.red(`Error: ${error.message}`));
      process.exit(1);
  }
}

main().catch(error => {
  console.error(chalk.red(`Unhandled error: ${error.message}`));
  process.exit(1);
});

File: scripts_n_helpers/printEndOfMakeInfo.py
#!/usr/bin/env python3

import json
import random
import string
import boto3
from termcolor import colored
import os
import argparse
import time

def random_password():
    char_code = random.randint(35, 47)
    part1 = ''.join(random.choices(string.ascii_lowercase + string.digits, k=8))
    part2 = ''.join(random.choices(string.ascii_uppercase + string.digits, k=8))
    return part1 + chr(char_code) + part2

# Set up command line argument parsing
parser = argparse.ArgumentParser(description='Generate Cognito user commands')
parser.add_argument('-e', '--email', help='Email address for the user')
args = parser.parse_args()

# Use command line email if provided, otherwise prompt
email_address = args.email if args.email else input("Please enter your email address: ").strip()

# Generate password
password = random_password()
password_output = f"Your randomly generated password is: {colored(password, 'yellow', 'on_black')}"

# Get AWS account ID
sts_client = boto3.client('sts', region_name='us-west-2')
account_id = sts_client.get_caller_identity()['Account']

# Read outputs file
with open('./outputs.json', 'r') as file:
    outputs = json.load(file)

chatbot_url_output = ""

# Process outputs
for key, value in outputs.items():
    if "-backend-" in key:
        continue
    elif "-" in key:
        for inner_key, inner_value in value.items():
            if inner_key.startswith("WebsiteUserInterfaceDomainName"):
                chatbot_url_output = f"DeepRacer Chatbot URL: {colored(inner_value, 'blue')}"
                print(chatbot_url_output)
            elif inner_key.startswith("ExportsOutputRefAuthenticationUserPool"):
                user_pool_id = inner_value
                cognito_client = boto3.client('cognito-idp', region_name='us-west-2')
                
                print("\nStarting user creation process...")
                
                # Step 1: Create user
                print("\nExecuting: Creating user")
                try:
                    cognito_client.admin_create_user(
                        UserPoolId=user_pool_id,
                        Username=email_address,
                        MessageAction='SUPPRESS'  # Suppress sending the temporary password
                    )
                    print(colored("Success!", "green"))
                except cognito_client.exceptions.UsernameExistsException:
                    print(colored("Note: User already exists", "yellow"))
                except Exception as e:
                    print(colored(f"✗ Error creating user: {str(e)}", "red"))
                    exit(1)  # Exit if we can't create/verify user
                
                time.sleep(2)  # Wait for user creation to propagate
                
                # Step 2: Add user to group
                print("\nExecuting: Adding user to group")
                try:
                    cognito_client.admin_add_user_to_group(
                        UserPoolId=user_pool_id,
                        Username=email_address,
                        GroupName='Users'
                    )
                    print(colored("Success!", "green"))
                except Exception as e:
                    print(colored(f"✗ Error adding user to group: {str(e)}", "red"))
                    exit(1)  # Exit if we can't add to group
                
                time.sleep(1)
                
                # Step 3: Set permanent password
                print("\nExecuting: Setting user password")
                try:
                    cognito_client.admin_set_user_password(
                        UserPoolId=user_pool_id,
                        Username=email_address,
                        Password=password,
                        Permanent=True
                    )
                    print(colored("Success!", "green"))
                    print('')
                    print(password_output)
                except Exception as e:
                    print(colored(f"✗ Error setting password: {str(e)}", "red"))
                    exit(1)  # Exit if we can't set password

print()
